src/caffeine/allocation_s.f90

```diff
@@ -1,25 +1,154 @@
+    ! TODO: determining the size of the handle and where the coarray begins
+    !       becomes a bit more complicated if we don't allocate space for
+    !       15 cobounds
+    type(c_ptr) :: whole_block
+    integer(c_size_t) :: handle_size, coarray_size, total_size
+    type(handle_data) :: unused
+    type(handle_data), pointer :: unused2(:)
 
-    allocated_memory = caf_allocate( &
-      product(ubounds-lbounds+1)*element_length, size(ucobounds), lcobounds, ucobounds, final_func, coarray_handle%ptr)
+    coarray_size = product(ubounds-lbounds+1)*element_length
+    handle_size = c_sizeof(unused)
+    total_size = handle_size + coarray_size
+
+    ! TODO: have only "team leader" perform allocation
+    !       and "broadcast" offset as part of synchronization
+    whole_block = caf_allocate(current_team%heap, total_size)
+    call prif_sync_all
+
+    call c_f_pointer(whole_block, coarray_handle%info)
+    call c_f_pointer(whole_block, unused2, [2])
+
+    coarray_handle%info%coarray_data = c_loc(unused2(2))
+    coarray_handle%info%corank = size(lcobounds)
+    coarray_handle%info%coarray_size = coarray_size
+    coarray_handle%info%final_func = final_func
+    coarray_handle%info%cobounds(1:size(lcobounds))%lcobound = lcobounds
+    coarray_handle%info%cobounds(1:size(lcobounds))%ucobound = ucobounds
+    call add_to_team_list(current_team, coarray_handle)
+
+    allocated_memory = coarray_handle%info%coarray_data
   end procedure
 
   module procedure prif_allocate_non_symmetric
   end procedure
 
   module procedure prif_deallocate
+    ! gfortran is yelling that this isn't valid for bind(C)
+    ! https://gcc.gnu.org/bugzilla/show_bug.cgi?id=113338
+    ! abstract interface
+    !   subroutine coarray_cleanup_i(handle, stat, errmsg) bind(C)
+    !     import c_int, prif_coarray_handle
+    !     implicit none
+    !     type(prif_coarray_handle), pointer, intent(in) :: handle
+    !     integer(c_int), intent(out) :: stat
+    !     character(len=:), intent(out), allocatable :: errmsg
+    !   end subroutine
+    ! end interface
+    integer :: i, num_handles
+    integer(c_int) :: local_stat
+    character(len=:), allocatable :: local_errmsg
+    ! procedure(coarray_cleanup_i), pointer :: coarray_cleanup
+    character(len=*), parameter :: unallocated_message = "Attempted to deallocate unallocated coarray"
+
+    call prif_sync_all ! Need to ensure we don't deallocate anything till everyone gets here
+    num_handles = size(coarray_handles)
+    if (.not. all([(associated(coarray_handles(i)%info), i = 1, num_handles)])) then
+      if (present(stat)) then
+        stat = 1 ! TODO: decide what our stat codes should be
+        if (present(errmsg)) then
+          errmsg = unallocated_message
+        else if (present(errmsg_alloc)) then
+          errmsg_alloc = unallocated_message
+        end if
+        return
+      else
+        call prif_error_stop(.false._c_bool, stop_code_char=unallocated_message)
+      end if
+    end if
```

Comment:
```suggestion
    end if
    ! TODO: invoke finalizers from coarray_handles(:)%info%final_func
```
I'd like to hear more about why Fortran is rejecting the commented-out code, but that discussion can take place post-merge. 

We can always invoke the finalizer from the C-side if necessary to circumvent any Fortran-side stupidity.

Reply:
It's a bug in gfortran. I've submitted a bug-report. There is a subtle change we'll need to make to the interface of `prif_deallocate` though.